triton/stream.py

```diff
@@ -253,13 +260,227 @@ def checkpoint(self):
 
 
 class Stream(object):
+    __metaclass__ = abc.ABCMeta
+
+
+    @abc.abstractmethod
+    def put(self, **record):
+        pass
+
+
+    @abc.abstractmethod
+    def put_many(self, records):
+        pass
+
+
+    def encode(self, record):
+        try:
+            data = msgpack.packb(record)
+        except TypeError:
+            # If we fail to serialize our context, we can try again with an
+            # enhanced packer (it's slower though)
+            data = msgpack.packb(record, default=msgpack_encode_default)
+
+        return data
+
+
+    def decode(self, record_blob):
+        return msgpack.unpackb(record_blob)
+
+
+    def __iter__(self):
+        return self.build_iterator_from_latest()
+
+
+    @abc.abstractmethod
+    def build_iterator_from_latest(self, **kwargs):
+        pass
+
+
+    @abc.abstractmethod
+    def build_iterator_from_checkpoint(self, checkpoint):
+        pass
+
+class CompositeStream(Stream):
+
+
+    class Iterator(object):
+
+
+        def __init__(self, streams, checkpoints=[]):
+            if checkpoints == []:
+                self.iterators = [stream.build_iterator_from_latest() for stream in streams]
+            else:
+                self.iterators = [stream.build_iterator_from_checkpoint(checkpoint)
+                                    for stream, checkpoint in zip(streams, checkpoints)]
+
+            self.iterator = itertools.izip_longest(*tuple(self.iterators), fillvalue=None)
+
+
+        def __iter__(self):
+            return self.iterator
+
+
+    def __init__(self, streams):
+        self.streams = streams
+
+
+    def put(self, **record):
+        return [stream.put(**record) for stream in self.streams]
+
+
+    def put_many(self, records):
+        return [stream.put_many(records) for stream in self.streams]
+
+
+    def build_iterator_from_latest(self):
+        return CompositeStream.Iterator(self.streams)
+
+
+    def build_iterator_from_checkpoint(self, checkpoints):
+        return CompositeStream.Iterator(self.streams, checkpoints)
+
+
+class GCPStream(Stream):
+
+
+    # Pubsub limits as defined https://cloud.google.com/pubsub/quotas.
+    BATCH_MAX_MSGS = 1000
+
+    class Iterator(object):
+
+
+        def __init__(self, stream, subscription_id=None):
+            self.stream = stream
+            self.subscription_id = subscription_id or 'trition-{}'.format(uuid.uuid4().hex)
+            self.subscription = stream.topic.subscription(self.subscription_id)
+            if subscription_id is None:
+                # If the subscription_id is provided, then
+                # it is assumed we are resuming from a previously
+                # active, named subscription.
+                self.subscription.create()
+
+            self.prefetch = []
+            self.prefetch_ack_ids = []
+            self.pending_acks = []
+            self.first_next = True
+
+
+        def __iter__(self):
+            return self
+
+
+        def next(self):
+            return self.__next__()
```

Comment:
You could also just do `next = __next__` at the def level right after the `__next__` method is defined